File: crates/gust-core/src/middleware/circuit_breaker.rs
//! Circuit Breaker Middleware
//!
//! Fault tolerance pattern for handling failures gracefully.
//! Implements the circuit breaker pattern with three states:
//! - Closed: Normal operation, requests pass through
//! - Open: Circuit tripped, requests fail fast
//! - Half-Open: Testing if service recovered

use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
use std::sync::RwLock;
use std::time::{Duration, Instant};

/// Circuit breaker state
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitState {
    /// Normal operation - requests pass through
    Closed,
    /// Circuit tripped - requests fail immediately
    Open,
    /// Testing recovery - limited requests allowed
    HalfOpen,
}

impl CircuitState {
    pub fn as_str(&self) -> &'static str {
        match self {
            CircuitState::Closed => "closed",
            CircuitState::Open => "open",
            CircuitState::HalfOpen => "half-open",
        }
    }
}

/// Circuit breaker statistics
#[derive(Debug, Clone)]
pub struct CircuitStats {
    pub state: CircuitState,
    pub failures: u32,
    pub successes: u32,
    pub total_requests: u64,
    pub total_failures: u64,
    pub total_successes: u64,
    pub last_failure_ms: Option<u64>,
    pub last_success_ms: Option<u64>,
}

/// Circuit breaker configuration
#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    /// Failure threshold to open circuit (default: 5)
    pub failure_threshold: u32,
    /// Success threshold to close circuit from half-open (default: 2)
    pub success_threshold: u32,
    /// Time before trying again (default: 30s)
    pub reset_timeout: Duration,
    /// Time window for counting failures (default: 60s)
    pub failure_window: Duration,
    /// Request timeout (default: 10s)
    pub timeout: Duration,
    /// Name for monitoring
    pub name: String,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_threshold: 5,
            success_threshold: 2,
            reset_timeout: Duration::from_secs(30),
            failure_window: Duration::from_secs(60),
            timeout: Duration::from_secs(10),
            name: "default".to_string(),
        }
    }
}

impl CircuitBreakerConfig {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            ..Default::default()
        }
    }

    pub fn failure_threshold(mut self, threshold: u32) -> Self {
        self.failure_threshold = threshold;
        self
    }

    pub fn success_threshold(mut self, threshold: u32) -> Self {
        self.success_threshold = threshold;
        self
    }

    pub fn reset_timeout(mut self, timeout: Duration) -> Self {
        self.reset_timeout = timeout;
        self
    }

    pub fn failure_window(mut self, window: Duration) -> Self {
        self.failure_window = window;
        self
    }

    pub fn timeout(mut self, timeout: Duration) -> Self {
        self.timeout = timeout;
        self
    }
}

/// Circuit breaker implementation
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    state: RwLock<CircuitState>,
    /// Timestamps of recent failures (within failure window)
    failures: RwLock<Vec<Instant>>,
    /// Successes in half-open state
    half_open_successes: AtomicU32,
    /// Next attempt time (when circuit is open)
    next_attempt: RwLock<Option<Instant>>,
    /// Statistics
    total_requests: AtomicU64,
    total_failures: AtomicU64,
    total_successes: AtomicU64,
    last_failure: RwLock<Option<Instant>>,
    last_success: RwLock<Option<Instant>>,
    /// Creation time for relative timestamps
    created_at: Instant,
}

impl CircuitBreaker {
    /// Create a new circuit breaker
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            config,
            state: RwLock::new(CircuitState::Closed),
            failures: RwLock::new(Vec::new()),
            half_open_successes: AtomicU32::new(0),
            next_attempt: RwLock::new(None),
            total_requests: AtomicU64::new(0),
            total_failures: AtomicU64::new(0),
            total_successes: AtomicU64::new(0),
            last_failure: RwLock::new(None),
            last_success: RwLock::new(None),
            created_at: Instant::now(),
        }
    }

    /// Get current state
    pub fn state(&self) -> CircuitState {
        *self.state.read().unwrap()
    }

    /// Get circuit breaker name
    pub fn name(&self) -> &str {
        &self.config.name
    }

    /// Get timeout duration
    pub fn timeout(&self) -> Duration {
        self.config.timeout
    }

    /// Get statistics
    pub fn stats(&self) -> CircuitStats {
        let failures = self.failures.read().unwrap();
        let last_failure = self.last_failure.read().unwrap();
        let last_success = self.last_success.read().unwrap();

        CircuitStats {
            state: self.state(),
            failures: failures.len() as u32,
            successes: self.half_open_successes.load(Ordering::Relaxed),
            total_requests: self.total_requests.load(Ordering::Relaxed),
            total_failures: self.total_failures.load(Ordering::Relaxed),
            total_successes: self.total_successes.load(Ordering::Relaxed),
            last_failure_ms: last_failure.map(|t| t.duration_since(self.created_at).as_millis() as u64),
            last_success_ms: last_success.map(|t| t.duration_since(self.created_at).as_millis() as u64),
        }
    }

    /// Check if request can proceed
    pub fn can_request(&self) -> bool {
        let state = self.state();
        match state {
            CircuitState::Closed => true,
            CircuitState::Open => {
                // Check if we can try again
                let next = self.next_attempt.read().unwrap();
                if let Some(next_time) = *next {
                    if Instant::now() >= next_time {
                        drop(next);
                        self.to_half_open();
                        return true;
                    }
                }
                false
            }
            CircuitState::HalfOpen => true,
        }
    }

    /// Record a successful request
    pub fn record_success(&self) {
        self.total_requests.fetch_add(1, Ordering::Relaxed);
        self.total_successes.fetch_add(1, Ordering::Relaxed);
        *self.last_success.write().unwrap() = Some(Instant::now());

        let state = self.state();
        if state == CircuitState::HalfOpen {
            let successes = self.half_open_successes.fetch_add(1, Ordering::Relaxed) + 1;
            if successes >= self.config.success_threshold {
                self.to_closed();
            }
        }
    }

    /// Record a failed request
    pub fn record_failure(&self) {
        self.total_requests.fetch_add(1, Ordering::Relaxed);
        self.total_failures.fetch_add(1, Ordering::Relaxed);
        *self.last_failure.write().unwrap() = Some(Instant::now());

        let state = self.state();
        match state {
            CircuitState::HalfOpen => {
                // Any failure in half-open immediately opens circuit
                self.to_open();
            }
            CircuitState::Closed => {
                // Add failure timestamp
                let now = Instant::now();
                let mut failures = self.failures.write().unwrap();
                failures.push(now);

                // Remove old failures outside window
                let window_start = now - self.config.failure_window;
                failures.retain(|t| *t > window_start);

                if failures.len() as u32 >= self.config.failure_threshold {
                    drop(failures);
                    self.to_open();
                }
            }
            CircuitState::Open => {
                // Already open, nothing to do
            }
        }
    }

    /// Force open the circuit
    pub fn open(&self) {
        self.to_open();
    }

    /// Force close the circuit
    pub fn close(&self) {
        self.to_closed();
    }

    /// Reset the circuit
    pub fn reset(&self) {
        *self.state.write().unwrap() = CircuitState::Closed;
        self.failures.write().unwrap().clear();
        self.half_open_successes.store(0, Ordering::Relaxed);
        *self.next_attempt.write().unwrap() = None;
    }

    fn to_open(&self) {
        let mut state = self.state.write().unwrap();
        if *state != CircuitState::Open {
            *state = CircuitState::Open;
            *self.next_attempt.write().unwrap() = Some(Instant::now() + self.config.reset_timeout);
            self.half_open_successes.store(0, Ordering::Relaxed);
        }
    }

    fn to_half_open(&self) {
        let mut state = self.state.write().unwrap();
        if *state != CircuitState::HalfOpen {
            *state = CircuitState::HalfOpen;
            self.half_open_successes.store(0, Ordering::Relaxed);
        }
    }

    fn to_closed(&self) {
        let mut state = self.state.write().unwrap();
        if *state != CircuitState::Closed {
            *state = CircuitState::Closed;
            self.failures.write().unwrap().clear();
            self.half_open_successes.store(0, Ordering::Relaxed);
        }
    }
}

/// Bulkhead configuration (concurrency limiter)
#[derive(Debug, Clone)]
pub struct BulkheadConfig {
    /// Maximum concurrent requests
    pub max_concurrent: u32,
    /// Maximum queue size
    pub max_queue: u32,
    /// Queue timeout
    pub queue_timeout: Duration,
}

impl Default for BulkheadConfig {
    fn default() -> Self {
        Self {
            max_concurrent: 10,
            max_queue: 100,
            queue_timeout: Duration::from_secs(30),
        }
    }
}

impl BulkheadConfig {
    pub fn new(max_concurrent: u32) -> Self {
        Self {
            max_concurrent,
            ..Default::default()
        }
    }

    pub fn max_queue(mut self, max: u32) -> Self {
        self.max_queue = max;
        self
    }

    pub fn queue_timeout(mut self, timeout: Duration) -> Self {
        self.queue_timeout = timeout;
        self
    }
}

/// Bulkhead (concurrency limiter)
pub struct Bulkhead {
    config: BulkheadConfig,
    running: AtomicU32,
    queued: AtomicU32,
}

impl Bulkhead {
    pub fn new(config: BulkheadConfig) -> Self {
        Self {
            config,
            running: AtomicU32::new(0),
            queued: AtomicU32::new(0),
        }
    }

    /// Try to acquire a slot
    /// Returns Ok(()) if acquired, Err(reason) if rejected
    pub fn try_acquire(&self) -> Result<BulkheadGuard<'_>, &'static str> {
        let current = self.running.load(Ordering::Relaxed);
        if current < self.config.max_concurrent {
            // Try to acquire
            if self.running.compare_exchange(
                current,
                current + 1,
                Ordering::Acquire,
                Ordering::Relaxed,
            ).is_ok() {
                return Ok(BulkheadGuard { bulkhead: self });
            }
        }

        // Check queue capacity
        let queued = self.queued.load(Ordering::Relaxed);
        if queued >= self.config.max_queue {
            return Err("Queue full");
        }

        Err("At capacity")
    }

    /// Get current running count
    pub fn running(&self) -> u32 {
        self.running.load(Ordering::Relaxed)
    }

    /// Get current queued count
    pub fn queued(&self) -> u32 {
        self.queued.load(Ordering::Relaxed)
    }
}

/// Guard that releases bulkhead slot on drop
pub struct BulkheadGuard<'a> {
    bulkhead: &'a Bulkhead,
}

impl<'a> Drop for BulkheadGuard<'a> {
    fn drop(&mut self) {
        self.bulkhead.running.fetch_sub(1, Ordering::Release);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circuit_breaker_closed() {
        let cb = CircuitBreaker::new(CircuitBreakerConfig::default());
        assert_eq!(cb.state(), CircuitState::Closed);
        assert!(cb.can_request());
    }

    #[test]
    fn test_circuit_breaker_opens_on_failures() {
        let config = CircuitBreakerConfig::default()
            .failure_threshold(3)
            .failure_window(Duration::from_secs(60));
        let cb = CircuitBreaker::new(config);

        // Record failures
        cb.record_failure();
        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Closed);

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);
        assert!(!cb.can_request());
    }

    #[test]
    fn test_circuit_breaker_success_resets() {
        let config = CircuitBreakerConfig::default()
            .failure_threshold(3)
            .success_threshold(2);
        let cb = CircuitBreaker::new(config);

        // Open the circuit
        cb.record_failure();
        cb.record_failure();
        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);

        // Force to half-open for testing
        cb.close();
        cb.open();

        // Reset and verify
        cb.reset();
        assert_eq!(cb.state(), CircuitState::Closed);
    }

    #[test]
    fn test_circuit_breaker_stats() {
        let cb = CircuitBreaker::new(CircuitBreakerConfig::new("test"));

        cb.record_success();
        cb.record_failure();

        let stats = cb.stats();
        assert_eq!(stats.total_requests, 2);
        assert_eq!(stats.total_successes, 1);
        assert_eq!(stats.total_failures, 1);
    }

    #[test]
    fn test_bulkhead_capacity() {
        let bulkhead = Bulkhead::new(BulkheadConfig::new(2));

        let _g1 = bulkhead.try_acquire().unwrap();
        let _g2 = bulkhead.try_acquire().unwrap();

        // Should be at capacity
        assert!(bulkhead.try_acquire().is_err());
        assert_eq!(bulkhead.running(), 2);
    }

    #[test]
    fn test_bulkhead_release() {
        let bulkhead = Bulkhead::new(BulkheadConfig::new(1));

        {
            let _guard = bulkhead.try_acquire().unwrap();
            assert_eq!(bulkhead.running(), 1);
        }

        // Guard dropped, should be able to acquire again
        assert_eq!(bulkhead.running(), 0);
        assert!(bulkhead.try_acquire().is_ok());
    }
}

File: packages/gust/src/circuitBreaker.ts
/**
 * Circuit Breaker
 * Fault tolerance pattern for handling failures gracefully
 *
 * Architecture:
 * - Uses native Rust implementation when available (via gust-napi)
 * - Falls back to pure TypeScript for edge/serverless environments
 */

import { EventEmitter } from 'node:events'
import type { Handler, ServerResponse, Wrapper } from '@sylphx/gust-core'
import { response } from '@sylphx/gust-core'
import type { Context } from './context'
import {
	createNativeBulkhead,
	createNativeCircuitBreaker,
	type NativeBulkheadConfig,
	type NativeCircuitBreaker,
	type NativeCircuitBreakerConfig,
} from './native'

// ============================================================================
// Types
// ============================================================================

export type CircuitState = 'closed' | 'open' | 'half-open'

export type CircuitBreakerOptions = {
	/** Failure threshold to open circuit (default: 5) */
	readonly failureThreshold?: number
	/** Success threshold to close circuit (default: 2) */
	readonly successThreshold?: number
	/** Time in ms before trying again (default: 30s) */
	readonly resetTimeout?: number
	/** Time window for counting failures (default: 60s) */
	readonly failureWindow?: number
	/** Timeout for each request (default: 10s) */
	readonly timeout?: number
	/** Custom fallback response */
	readonly fallback?: (ctx: Context, error: Error) => ServerResponse | Promise<ServerResponse>
	/** Consider response a failure */
	readonly isFailure?: (res: ServerResponse) => boolean
	/** Name for this circuit (for monitoring) */
	readonly name?: string
	/** On state change callback */
	readonly onStateChange?: (state: CircuitState, name: string) => void
}

export type CircuitStats = {
	state: CircuitState
	failures: number
	successes: number
	lastFailure: number | null
	lastSuccess: number | null
	totalRequests: number
	totalFailures: number
	totalSuccesses: number
}

// ============================================================================
// Circuit Breaker Implementation
// ============================================================================

/**
 * Circuit Breaker
 *
 * Uses native Rust implementation when available for maximum performance.
 * Falls back to pure TypeScript for edge/serverless environments.
 */
export class CircuitBreaker extends EventEmitter {
	// Native implementation (when available)
	private native: NativeCircuitBreaker | null = null

	// Fallback TypeScript state
	private state: CircuitState = 'closed'
	private failures: number[] = [] // Timestamps of failures
	private successes = 0
	private lastFailure: number | null = null
	private lastSuccess: number | null = null
	private totalRequests = 0
	private totalFailures = 0
	private totalSuccesses = 0
	private nextAttempt = 0

	private readonly failureThreshold: number
	private readonly successThreshold: number
	private readonly resetTimeout: number
	private readonly failureWindow: number
	private readonly timeout: number
	private readonly name: string

	constructor(options: CircuitBreakerOptions = {}) {
		super()
		this.failureThreshold = options.failureThreshold ?? 5
		this.successThreshold = options.successThreshold ?? 2
		this.resetTimeout = options.resetTimeout ?? 30000
		this.failureWindow = options.failureWindow ?? 60000
		this.timeout = options.timeout ?? 10000
		this.name = options.name ?? 'default'

		// Try to use native implementation
		const nativeConfig: NativeCircuitBreakerConfig = {
			failureThreshold: this.failureThreshold,
			successThreshold: this.successThreshold,
			resetTimeoutMs: this.resetTimeout,
			failureWindowMs: this.failureWindow,
			timeoutMs: this.timeout,
			name: this.name,
		}
		this.native = createNativeCircuitBreaker(nativeConfig)
	}

	/** Check if using native implementation */
	get isNative(): boolean {
		return this.native !== null
	}

	/**
	 * Get current state
	 */
	getState(): CircuitState {
		if (this.native) {
			return this.native.state() as CircuitState
		}
		return this.state
	}

	/**
	 * Get statistics
	 */
	getStats(): CircuitStats {
		if (this.native) {
			const stats = this.native.stats()
			return {
				state: stats.state as CircuitState,
				failures: stats.failures,
				successes: stats.successes,
				lastFailure: this.lastFailure,
				lastSuccess: this.lastSuccess,
				totalRequests: stats.totalRequests,
				totalFailures: stats.totalFailures,
				totalSuccesses: stats.totalSuccesses,
			}
		}
		return {
			state: this.state,
			failures: this.failures.length,
			successes: this.successes,
			lastFailure: this.lastFailure,
			lastSuccess: this.lastSuccess,
			totalRequests: this.totalRequests,
			totalFailures: this.totalFailures,
			totalSuccesses: this.totalSuccesses,
		}
	}

	/**
	 * Check if request can proceed
	 */
	canRequest(): boolean {
		if (this.native) {
			return this.native.canRequest()
		}

		if (this.state === 'closed') return true
		if (this.state === 'open') {
			// Check if we can try again
			if (Date.now() >= this.nextAttempt) {
				this.toHalfOpen()
				return true
			}
			return false
		}
		// Half-open: allow one request
		return true
	}

	/**
	 * Record success
	 */
	recordSuccess(): void {
		if (this.native) {
			this.native.recordSuccess()
			this.lastSuccess = Date.now()
			return
		}

		this.totalRequests++
		this.totalSuccesses++
		this.lastSuccess = Date.now()

		if (this.state === 'half-open') {
			this.successes++
			if (this.successes >= this.successThreshold) {
				this.toClosed()
			}
		}
	}

	/**
	 * Record failure
	 */
	recordFailure(): void {
		if (this.native) {
			this.native.recordFailure()
			this.lastFailure = Date.now()
			return
		}

		this.totalRequests++
		this.totalFailures++
		this.lastFailure = Date.now()

		if (this.state === 'half-open') {
			this.toOpen()
			return
		}

		if (this.state === 'closed') {
			// Add failure timestamp
			this.failures.push(Date.now())

			// Remove old failures outside window
			const windowStart = Date.now() - this.failureWindow
			this.failures = this.failures.filter((t) => t > windowStart)

			if (this.failures.length >= this.failureThreshold) {
				this.toOpen()
			}
		}
	}

	/**
	 * Force open the circuit
	 */
	open(): void {
		this.toOpen()
	}

	/**
	 * Force close the circuit
	 */
	close(): void {
		this.toClosed()
	}

	/**
	 * Reset the circuit
	 */
	reset(): void {
		if (this.native) {
			this.native.reset()
			this.emit('reset', this.name)
			return
		}

		this.state = 'closed'
		this.failures = []
		this.successes = 0
		this.nextAttempt = 0
		this.emit('reset', this.name)
	}

	private toOpen(): void {
		if (this.state !== 'open') {
			this.state = 'open'
			this.nextAttempt = Date.now() + this.resetTimeout
			this.successes = 0
			this.emit('open', this.name)
		}
	}

	private toHalfOpen(): void {
		if (this.state !== 'half-open') {
			this.state = 'half-open'
			this.successes = 0
			this.emit('half-open', this.name)
		}
	}

	private toClosed(): void {
		if (this.state !== 'closed') {
			this.state = 'closed'
			this.failures = []
			this.successes = 0
			this.emit('close', this.name)
		}
	}
}

// ============================================================================
// Middleware
// ============================================================================

/**
 * Circuit breaker middleware
 */
export const circuitBreaker = (options: CircuitBreakerOptions = {}): Wrapper<Context> => {
	const {
		timeout = 10000,
		fallback,
		isFailure = (res) => res.status >= 500,
		onStateChange,
	} = options

	const breaker = new CircuitBreaker(options)

	// Set up state change listener
	if (onStateChange) {
		breaker.on('open', (name) => onStateChange('open', name))
		breaker.on('half-open', (name) => onStateChange('half-open', name))
		breaker.on('close', (name) => onStateChange('closed', name))
	}

	const defaultFallback = (): ServerResponse =>
		response(
			JSON.stringify({
				error: 'Service Unavailable',
				message: 'Circuit breaker is open',
			}),
			{
				status: 503,
				headers: {
					'content-type': 'application/json',
					'retry-after': String(Math.ceil((options.resetTimeout ?? 30000) / 1000)),
				},
			}
		)

	return (handler: Handler<Context>): Handler<Context> => {
		return async (ctx: Context): Promise<ServerResponse> => {
			// Check if circuit allows request
			if (!breaker.canRequest()) {
				if (fallback) {
					return fallback(ctx, new Error('Circuit breaker is open'))
				}
				return defaultFallback()
			}

			try {
				// Execute with timeout
				const result = await Promise.race([
					handler(ctx),
					new Promise<ServerResponse>((_, reject) =>
						setTimeout(() => reject(new Error('Request timeout')), timeout)
					),
				])

				// Check if response indicates failure
				if (isFailure(result)) {
					breaker.recordFailure()
				} else {
					breaker.recordSuccess()
				}

				return result
			} catch (error) {
				breaker.recordFailure()

				if (fallback) {
					return fallback(ctx, error as Error)
				}

				return response(
					JSON.stringify({
						error: 'Service Unavailable',
						message: (error as Error).message,
					}),
					{
						status: 503,
						headers: { 'content-type': 'application/json' },
					}
				)
			}
		}
	}
}

/**
 * Get circuit breaker for named service
 */
const breakers = new Map<string, CircuitBreaker>()

export const getCircuitBreaker = (
	name: string,
	options?: CircuitBreakerOptions
): CircuitBreaker => {
	let breaker = breakers.get(name)
	if (!breaker) {
		breaker = new CircuitBreaker({ ...options, name })
		breakers.set(name, breaker)
	}
	return breaker
}

/**
 * Wrap async function with circuit breaker
 */
export const withCircuitBreaker = <T extends (...args: unknown[]) => Promise<unknown>>(
	fn: T,
	options: CircuitBreakerOptions = {}
): T => {
	const breaker = new CircuitBreaker(options)

	return (async (...args: Parameters<T>): Promise<ReturnType<T>> => {
		if (!breaker.canRequest()) {
			throw new Error('Circuit breaker is open')
		}

		try {
			const result = await fn(...args)
			breaker.recordSuccess()
			return result as ReturnType<T>
		} catch (error) {
			breaker.recordFailure()
			throw error
		}
	}) as T
}

// ============================================================================
// Bulkhead (Concurrency Limiter)
// ============================================================================

export type BulkheadOptions = {
	/** Max concurrent requests (default: 10) */
	readonly maxConcurrent?: number
	/** Max queue size (default: 100) */
	readonly maxQueue?: number
	/** Queue timeout in ms (default: 30s) */
	readonly queueTimeout?: number
	/** Custom rejection response */
	readonly onReject?: (ctx: Context) => ServerResponse
}

/**
 * Bulkhead middleware (limit concurrency)
 *
 * Uses native Rust implementation when available for maximum performance.
 * Falls back to pure TypeScript for edge/serverless environments.
 */
export const bulkhead = (options: BulkheadOptions = {}): Wrapper<Context> => {
	const { maxConcurrent = 10, maxQueue = 100, queueTimeout = 30000, onReject } = options

	// Try to use native bulkhead
	const nativeConfig: NativeBulkheadConfig = {
		maxConcurrent,
		maxQueue,
		queueTimeoutMs: queueTimeout,
	}
	const nativeBulkhead = createNativeBulkhead(nativeConfig)

	// Fallback TypeScript state
	let running = 0
	const queue: Array<{
		resolve: () => void
		reject: (err: Error) => void
		timer: ReturnType<typeof setTimeout>
	}> = []

	const rejectResponse =
		onReject ??
		(() =>
			response(
				JSON.stringify({
					error: 'Service Unavailable',
					message: 'Too many concurrent requests',
				}),
				{
					status: 503,
					headers: {
						'content-type': 'application/json',
						'retry-after': '5',
					},
				}
			))

	const acquire = (): Promise<void> => {
		// Use native if available
		if (nativeBulkhead) {
			if (nativeBulkhead.tryAcquire()) {
				return Promise.resolve()
			}
			return Promise.reject(new Error('Queue full'))
		}

		// Fallback TypeScript implementation
		if (running < maxConcurrent) {
			running++
			return Promise.resolve()
		}

		if (queue.length >= maxQueue) {
			return Promise.reject(new Error('Queue full'))
		}

		return new Promise((resolve, reject) => {
			const timer = setTimeout(() => {
				const idx = queue.findIndex((q) => q.resolve === resolve)
				if (idx !== -1) queue.splice(idx, 1)
				reject(new Error('Queue timeout'))
			}, queueTimeout)

			queue.push({ resolve, reject, timer })
		})
	}

	const release = (): void => {
		// Native bulkhead handles release internally via RAII
		if (nativeBulkhead) return

		running--
		const next = queue.shift()
		if (next) {
			clearTimeout(next.timer)
			running++
			next.resolve()
		}
	}

	return (handler: Handler<Context>): Handler<Context> => {
		return async (ctx: Context): Promise<ServerResponse> => {
			try {
				await acquire()
			} catch {
				return rejectResponse(ctx)
			}

			try {
				return await handler(ctx)
			} finally {
				release()
			}
		}
	}
}
